openfisca_core/formulas.py

```diff
@@ -834,110 +823,71 @@ def new_filled_column(
-                            and reference_dated_formula_class['start_instant'] < dated_formula_class[
-                                'stop_instant']:
-                        reference_dated_formula_class['start_instant'] = dated_formula_class['stop_instant'].offset(
-                            1, 'day')
-                    if reference_dated_formula_class['stop_instant'] > dated_formula_class['start_instant'] \
-                            and reference_dated_formula_class['stop_instant'] <= dated_formula_class[
-                                'stop_instant']:
-                        reference_dated_formula_class['stop_instant'] = dated_formula_class['start_instant'].offset(
-                            -1, 'day')
-                    if reference_dated_formula_class['start_instant'] > reference_dated_formula_class[
-                            'stop_instant']:
-                        break
-                else:
-                    dated_formulas_class.append(reference_dated_formula_class)
-            dated_formulas_class.sort(key = lambda dated_formula_class: dated_formula_class['start_instant'])
+    # Turn function into a dated function
+    formula_name_prefix = 'formula'
+    formula_name_separator = '_'
+    dated_formulas_class = []
+    for function_name, function in specific_attributes.copy().iteritems():
+        if not function_name.startswith(formula_name_prefix):
+            # Current attribute isn't a formula
+            continue
+
+        formula_start_date = extract_formula_date_from_name(function_name, formula_name_prefix, formula_name_separator)
+        if end is not None:
+            end_date = get_datetime_date(name, end)
+            assert end_date >= formula_start_date, \
+                'You declared that "{}" ends on "{}", but you wrote a formula to calculate it from "{}" ({}). The "end" attribute of a variable must be posterior to the start dates of all its formulas.'.format(name, end_date, formula_start_date, function_name)
+        dated_formula_class_attributes = formula_class_attributes.copy()
+        dated_formula_class_attributes['formula'] = function
+        dated_formula_class = type(name.encode('utf-8'), (Formula,), dated_formula_class_attributes)
+
+        del specific_attributes[function_name]
+        dated_formulas_class.append(dict(
+            formula_class = dated_formula_class,
+            start_instant = periods.instant(formula_start_date),
+            ))
 
-        formula_class_attributes['dated_formulas_class'] = dated_formulas_class
-    else:
-        assert issubclass(formula_class, SimpleFormula), formula_class
+    # Sort dated formulas by start instant.
+    dated_formulas_class.sort(key = lambda dated_formula_class: dated_formula_class['start_instant'])
+
+    # Add dated formulas defined in (optional) reference column when they are not overridden by new dated
+    # formulas.
+    if reference_column is not None and issubclass(reference_column.formula_class, Formula):
```

Comment:
`issubclass(reference_column.formula_class, Formula)` is true by construction. Testing it is useless.